### PR TITLE
Add unit tests for spinw creation

### PR DESCRIPTION
**Pull request by rebeccafair**
This adds some tests for spinw object creation from different sources (might as well start with something easy...). The constructor tests aren't 100% complete, but I thought it would be good to get a review early in case my test structure is completely wrong. Summary of changes:

* Moved unit tests/systems tests into separate directories to keep things organised as we add more tests
* Added tests for creating a default spinw object, from a struct, from another spinw object, from a figure and from a file
* In adding tests for creation from a figure found a bug where the figure tag `'sw_crystal'` is searched for, this may be legacy as the actual tag for spinw structure figures is `'swplot'`. I refactored this bit of code anyway to just check for the spinw object rather than looking at the tag. There is another reference to `sw_crystal` in `sw_plotspec` but that can be sorted when we unit test that file
* In adding tests for creation from cif/fst files found some issues #82, #83. As a temporary fix for #83 improved the error message shown. There were also a few issues with reading files with tabs, and where comments have a ? in them, these have been fixed.

One thing, is there a better way to parametrise tests than putting them at the top of the file? Having the parameters so far away from the tests themselves might get annoying if we have a lot of tests...

### CODE REVIEW

**Review by mducle — APPROVED** [review]
Looks good to me :+1:

I think the test parameters have to be in a `properties (TestParameter)` block but that block does not have to be at the top so you can move it to just above the `methods (Test)` section even if this kind of goes against the "standard" formating...

An alternative is to just declare the property names in the `properties (TestParameter)` block but then to create a constructor which populates their values, maybe something like:

```matlab
    properties (TestParameter)
        spinw_struct_input;
        spinw_figure_input;
        spinw_file_input;
    end
    methods
        function obj = unittest_spinw()
                     spinw_struct_input = { ...
            {struct('lattice', struct('angle', [pi, pi, (2*pi)/3], ...
                                     'lat_const', [2, 2, 4])),
             'spinw_from_struct_lat224.mat'}, ...
            {struct('lattice', struct('angle', [pi; pi; (2*pi)/3], ...
                                     'lat_const', [2; 2; 4])), ...
             'spinw_from_struct_lat224.mat'}};
        end
    end
```

(I've not tested this - it might not work ;)

Finally, could you add an extra parameter "IncludingSubfolders" in the code coverage call in run_tests.m: `coverage_plugin = CodeCoveragePlugin.forFolder(cov_dirs{i}, 'Producing', reportFormat, 'IncludingSubfolders', true);` so it covers the package folders in `swfiles`, please?

File: +sw_tests/+unit_tests/unittest_spinw.m [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            testCase.verify_obj(expected_spinw, actual_spinw);
+            close(figure);
+        end
+        function test_spinw_from_incorrect_figure(testCase)
+            fig = figure('visible', 'off');
+            testCase.verifyError(@() spinw(fig), 'spinw:spinw:WrongInput');
+            close(fig);
+        end
+        function test_spinw_from_file(testCase, spinw_file_input)
+            fname = fullfile(testCase.get_unit_test_dir(), 'cifs', spinw_file_input{1});
```
Comment (RichardWaiteSTFC) [inline +sw_tests/+unit_tests/unittest_spinw.m:106]: I agree having the input structs at the top means you have to keep scrolling up/down to read the input. Is there any reason why we need to have the input as a property of the class, and not hard-code it in the call to the function? For example
`fname = fullfile(testCase.get_unit_test_dir(), 'cifs', 'YFeO3_mcphase.cif');`
if two tests were always intended to use the same file we could then make the filename a property instead of the whole input? Maybe that's poor form?

File: +sw_tests/+unit_tests/unittest_spinw.m [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                expected_value = expected_obj.(field{:});
+                actual_value = actual_obj.(field{:});
+                if isstruct(expected_value)
+                    testCase.verify_obj(expected_value, actual_value, rel_tol, abs_tol);
+                else
+                    testCase.verify_val(expected_value, actual_value, rel_tol, abs_tol, field{:});
+                end
+            end
+        end
+        function verify_val(testCase, expected_val, actual_val, rel_tol, abs_tol, field)
```
Comment (RichardWaiteSTFC) [inline +sw_tests/+unit_tests/unittest_spinw.m:55]: Feels like `verify_val` (and perhaps `verify_obj`) should be a utility function much like `mock_function`  - I can imagine wanting to use these in future unit tests.

File: +sw_tests/+unit_tests/unittest_spinw.m [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        function obj = load_spinw(testCase, filename)
+            obj = load(fullfile(testCase.get_unit_test_dir(), 'spinw', filename));
+            obj = obj.data;
+        end
+        function obj = load_figure(testCase, filename)
+            path = fullfile(testCase.get_unit_test_dir(), 'Figure', filename);
+            obj = openfig(path, 'invisible');
+        end
+        function verify_obj(testCase, expected_obj, actual_obj, rel_tol, abs_tol)
+            if nargin < 5
```
Comment (RichardWaiteSTFC) [inline +sw_tests/+unit_tests/unittest_spinw.m:34]: Atm `verify_val` is only called from `verify_obj`, so we can probably just have the default argument values for tolerance s in `verify_obj` and avoid some duplication (though in another comment I suggest putting this as a separate utility function)

File: +sw_tests/+unit_tests/unittest_spinw.m [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                abs_tol = 0;
+            end
+            if nargin < 4
+                rel_tol = 1e-10;
+            end
+            all_fieldnames = union(fieldnames(expected_obj), fieldnames(actual_obj));
```
Comment (RichardWaiteSTFC) [inline +sw_tests/+unit_tests/unittest_spinw.m:40]: Here should we not add an additional assertion that  f `expected_obj` and `actual_obj` have different numbers of fieldnames they are not equal? We can also assert that the objects are of the same class type?

**Review by RichardWaiteSTFC — CHANGES_REQUESTED** [review]
Looks good, I have a few comments, some might be overkill given there are only a couple of unit tests atm!

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 3 above)
Comment (RichardWaiteSTFC) [reply +sw_tests/+unit_tests/unittest_spinw.m:34]: It looks like MATLAB testing library has a struct comparator class, would that do an equivalent job to `verify_obj1 in this case?
https://uk.mathworks.com/help/matlab/ref/matlab.unittest.constraints.structcomparator-class.html

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 1 above)
Comment (rebeccafair) [reply +sw_tests/+unit_tests/unittest_spinw.m:106]: A test is run for every entry in each `TestParameter` e.g. `spinw_figure_input` has length 2 so `test_spinw_from_figure` is called twice, once with `{'default_structure.fig', 'spinw_default.mat'}` and once with `{'afm_chain_spec.fig', 'spinw_afm_chain.mat'}`. If we hard code the values we'd have to write 2 tests and end up duplicating a lot more code. https://uk.mathworks.com/help/matlab/matlab_prog/create-basic-parameterized-test.html

Repeated filenames aren't made a property as I'd like to be able to add more test cases easily (by just adding another entry in `spinw_figure_input`) if we find we need to cover more edge cases for example.

Sorry if I've misunderstood what you meant here! But I do agree the parameters being all at the top is not so good... in Pytest they're just above the function and it's a lot easier to read, so maybe there's a better Matlab-style way to doing this as currently I'm just using my Pytest brain.

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 2 above)
Comment (rebeccafair) [reply +sw_tests/+unit_tests/unittest_spinw.m:55]: Yes! I just put them here for now since we don't have any other test files yet, but I can move them out as part of this PR if you want to use them.

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 1 above)
Comment (RichardWaiteSTFC) [reply +sw_tests/+unit_tests/unittest_spinw.m:106]: Ah I see thanks, sorry I'm unfamiliar with MATLAB testing framework. In mantid it is normally preferred to have smaller similar tests so that it is obvious which use case breaks the test - e.g.
`test_spinw_figure_import_with_default_fig`
`test_spinw_figure_import_with_afm_chain`
with a helper function to handle the shared code between the two - but maybe MATLAB makes this difficult to do this in the same file! Whatever works, I'm happy as is, now I understand how the arguments are being passed!

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 2 above)
Comment (RichardWaiteSTFC) [reply +sw_tests/+unit_tests/unittest_spinw.m:55]: If you don't mind I think it makes sense to do it in this PR to avoid merge conflicts in subsequent PRs as we'll both be working on unit tests and probably both need it as a utility function!

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 1 above)
Comment (rebeccafair) [reply +sw_tests/+unit_tests/unittest_spinw.m:106]: Ah that is another valid way of doing things! You've got me looking up the advantages and disadvantages of parametrised tests now, that would be one option to avoid the inputs being so far away from the actual test...

File: +sw_tests/+unit_tests/unittest_spinw.m (hunk 4 above)
Comment (rebeccafair) [reply +sw_tests/+unit_tests/unittest_spinw.m:40]: Yeah good idea on testing the class type! And I wanted to test the combined fieldnames of both objects, if we just use the fields from the expected object, if the actual object has extra fields that have been added accidentally this wont be detected and the test will still pass.

**Review by RichardWaiteSTFC — APPROVED** [review]
Looks great with the test super class with the `verify_xxx` funcs - this keeps the actual spinw test file looking much cleaner!
